src/linear_algebra.jl

```diff
@@ -8,48 +8,220 @@ end
+            for i = k+1:minmn
+                p[i] = i
+            end
+            break
         end
+        rank = k
 
-        # swap
-        for j in 1:n
-            F[k, j], F[kp, j] = F[kp, j], F[k, j]
+        # swap rows
+        if k != kp
+            for j = 1:n
+                F[k, j], F[kp, j] = F[kp, j], F[k, j]
+            end
         end
 
-        for i in k+1:m
-            F[i, k] = F[i, k] / F[k, k]
+        # set values for L matrix
+        c = F[k, lead]
+        for i = k+1:m
+            F[i, k] = F[i, lead] / c
+            if lead != k
+                F[i, lead] = zero(Num)
+            end
         end
-        for j = k+1:n
-            for i in k+1:m
+
+        # substract the row from every other, traverse first by colums
+        # we start from lead+1 to avoid chaing the leading value on the column
+        for j = lead+1:n
+            for i = k+1:m
+                # this line occupies most of the time, distributed in the
+                # following methods
+                #   - `*(::Num, ::Num)` dynamic dispatch
+                #   - `-(::Num, ::Num)` dynamic dispatch
                 F[i, j] = F[i, j] - F[i, k] * F[k, j]
             end
         end
+
+        # advance the lead by one
+        lead = lead + 1
     end
+    return F, p, filter!(!iszero, leads), rank
+end
+function sym_lu(A; check=true)
+    F, p, leads, rank = _sym_lu(A)
+    info = rank == minimum(size(A)) ? 0 : rank
     check && LinearAlgebra.checknonsingular(info)
     LU(F, p, convert(LinearAlgebra.BlasInt, info))
 end
 
+# soft pivoted reduced row echelon form for extended matrix
+function sym_rref(A, b)
+    SINGULAR = typemax(Int)
+    m, n = size(A)
+    F = map(x->x isa Num ? x : Num(x), A)
+    b = map(x->x isa Num ? x : Num(x), b)
+    minmn = min(m, n)
+    lead = 1
+    leads = zeros(Int, minmn)
+    rank = 0
+    for k = 1:m
+        kp = k # pivot index
+
+        # search for the expression different from zero with the least terms
+        amin = SINGULAR
+        for j = lead:n
+            for i = k:m # search first by columns
+                absi = _iszero(F[i, j]) ? SINGULAR : nterms(F[i, j])
+                if absi < amin
+                    kp = i
+                    amin = absi
+                end
+            end
+            # break when pivot found
+            if amin != SINGULAR
+                lead = j
+                leads[k] = lead
+                break
+            end
+        end
+
+        # normally, here we would save the permutation in an array
+        # but this is not needed as we have the extended matrix
+
+        # break from function as the reduced echelon form has been reached
+        if amin == SINGULAR && !(amin isa Symbolic) && (amin isa Number)
+            break
+        end
+        rank = k
+
+        # swap rows, only needed to swap lead:end
+        if k != kp
+            for j = lead:n
+                F[k, j], F[kp, j] = F[kp, j], F[k, j]
```

Comment:
Rows of `b` are not being swapped, which results in incorrect throwing of inconsistent system for consistent systems. Adding 
```julia
b[k], b[kp] = b[kp, b[k]
```
fixes this